### PR TITLE
feat: add PMTK support

### PR DESCRIPTION
**Pull request by zhouhanseng**
Add prediction market tokens json file

learn more from https://github.com/DimensionDev/Maskbook/issues/2071

### CODE REVIEW

File: packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts [hunk 1]
```diff
@@ -32,6 +35,11 @@ const KEYWORK_ID_MAP: {
         CRU: 'crust-network',
```
Comment (Tedko) [inline packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts:35]: @guanbinrui do we still need to hard code $CRU given cmc already have themhttps://coinmarketcap.com/currencies/crust/ and CRU is on the top list ? 

Note: gecko haven't updated crust $CRU on top list though.

cc @zhouhanseng

**Comment by Tedko** [comment]
@guanbinrui need a review of this pr (Predict Mkt token func) and other merged pr such as #2076 ; and lets do a new release to chrome/ff Monday. I will ping Predict Mkt token guys for helping testing

File: packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts [hunk 2]
```diff
@@ -26,20 +24,14 @@ const KEYWORK_ID_MAP: {
 } = {
     [DataProvider.COIN_MARKET_CAP]: {
         UNI: '7083',
-        CRU: '6747',
-        CRUST: '6747',
```
Comment (zhouhanseng) [inline packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts:27]: Tested, we don't need to hard code $CRU for coinmarketcap.

File: packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts [hunk 3]
```diff
@@ -12,9 +12,7 @@ const KEYWORD_ALIAS_MAP: {
     [DataProvider.COIN_GECKO]: {
         NYFI: 'n0031',
     },
-    [DataProvider.UNISWAP]: {
```
Comment (zhouhanseng) [inline packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts]: My fault, there's nothing to hot fix with Uniswap, corrected.

File: packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts [hunk 4]
```diff
@@ -26,20 +24,12 @@ const KEYWORK_ID_MAP: {
 } = {
     [DataProvider.COIN_MARKET_CAP]: {
         UNI: '7083',
-        CRU: '6747',
-        CRUST: '6747',
         YAM: '7131', // v3
     },
     [DataProvider.COIN_GECKO]: {
         UNI: 'uniswap',
-        CRU: 'crust-network',
```
Comment (zhouhanseng) [inline packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts:32]: Tested, don't need to hard code $CRU for gecko neither.

**Comment by zhouhanseng** [comment]
Ready for reiview.

File: packages/maskbook/src/plugins/Trader/UI/trending/CoinMarketPanel.tsx [hunk 5]
```diff
@@ -24,7 +24,9 @@ export function CoinMarketPanel(props: CoinMarketPanelProps) {
         <div className={classes.root}>
             <CoinMetadataTable dataProvider={dataProvider} trending={trending} />
             <br />
-            <CoinMarketTable dataProvider={dataProvider} trending={trending} />
+            {dataProvider !== DataProvider.UNISWAP ? (
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/UI/trending/CoinMarketPanel.tsx:27]: Why not provide a Uniswap icon?

File: packages/maskbook/src/plugins/Trader/settings.ts [hunk 6]
```diff
@@ -82,8 +83,10 @@ export function getCurrentDataProviderGeneralSettings(dataProvider: DataProvider
             return coinGeckoSettings
         case DataProvider.COIN_MARKET_CAP:
             return coinMarketCapSettings
+        case DataProvider.UNISWAP:
+            return coinUniswapSettings
         default:
-            unreachable(dataProvider)
+            unreachable(dataProvider as never)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/settings.ts:89]: Should not use `as never`.

File: packages/maskbook/src/plugins/Trader/apis/trending/index.ts [hunk 7]
```diff
@@ -283,9 +286,31 @@ export async function getCoinTrendingByKeyword(
...
+    if (dataProvider === DataProvider.UNISWAP) {
+        return getCoinTrendingFromUniswap(currency, coin)
+    }
+
     return getCoinTrendingById(resolveCoinId(keyword_, dataProvider) ?? coin.id, currency, dataProvider)
 }
 
+export async function getCoinTrendingFromUniswap(currency: Currency, coin: Coin) {
+    const PMTk = new Token(ChainId.MAINNET, coin.eth_address!, 18, coin.symbol, coin.name)
+    const DAI = new Token(ChainId.MAINNET, '0x6b175474e89094c44da98b954eedeac495271d0f', 18, 'DAI', 'DAI')
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/apis/trending/index.ts:298]: You can get a DAI_ADDRESS from constants.

```
const DAI_ADDRESS = getConstant(CONSTANTS, 'DAI_ADDRESS')
```

File: packages/maskbook/src/plugins/Trader/settings.ts (hunk 6 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/settings.ts:89]: vscode complains, does this affect error print?

```
export function unreachable(val: never): never {
    console.error('Unhandled value: ', val)
    throw new Error('Unreachable case:' + val)
}
```

File: packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts [hunk 8]
```diff
@@ -45,25 +43,27 @@ const ID_ADDRESS_MAP: {
...
 }
 
 export function resolveAlias(keyword: string, dataProvider: DataProvider) {
     if (dataProvider === DataProvider.COIN_MARKET_CAP)
         return KEYWORD_ALIAS_MAP[DataProvider.COIN_MARKET_CAP][keyword.toUpperCase()] ?? keyword
     if (dataProvider === DataProvider.COIN_GECKO)
         return KEYWORD_ALIAS_MAP[DataProvider.COIN_GECKO][keyword.toUpperCase()] ?? keyword
-    unreachable(dataProvider)
+    if (dataProvider === DataProvider.UNISWAP) return keyword
+    unreachable(dataProvider as never)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts:55]: Should not use `as never` here.

File: packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts [hunk 9]
```diff
@@ -45,25 +43,27 @@ const ID_ADDRESS_MAP: {
...
 }
 
 export function resolveCoinId(keyword: string, dataProvider: DataProvider) {
     if (dataProvider === DataProvider.COIN_MARKET_CAP)
         return KEYWORK_ID_MAP[DataProvider.COIN_MARKET_CAP][keyword.toUpperCase()]
     if (dataProvider === DataProvider.COIN_GECKO) return KEYWORK_ID_MAP[DataProvider.COIN_GECKO][keyword.toUpperCase()]
-    unreachable(dataProvider)
+    unreachable(dataProvider as never)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts:62]: Should not use `as never` here.

File: packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts [hunk 10]
```diff
@@ -45,25 +43,27 @@ const ID_ADDRESS_MAP: {
...
 }
 
 export function resolveCoinAddress(id: string, dataProvider: DataProvider) {
     if (dataProvider === DataProvider.COIN_MARKET_CAP) return ID_ADDRESS_MAP[DataProvider.COIN_MARKET_CAP][id]
     if (dataProvider === DataProvider.COIN_GECKO) return ID_ADDRESS_MAP[DataProvider.COIN_GECKO][id]
-    unreachable(dataProvider)
+    unreachable(dataProvider as never)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts:68]: Should not use `as never` here.

Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts:68]: why?

File: packages/maskbook/src/plugins/Trader/apis/trending/index.ts [hunk 11]
```diff
@@ -9,6 +9,8 @@ import { resolveCoinId, resolveCoinAddress, resolveAlias } from './hotfix'
 import STOCKS_KEYWORDS from './stocks.json'
 import CASHTAG_KEYWORDS from './cashtag.json'
 import HASHTAG_KEYWORDS from './hashtag.json'
+import PREDICTION_MARKET_TOKENS from './prediction_market_tokens.json'
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/apis/trending/index.ts:12]: Please move this file to the `apis/unswap/index.ts`.

Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/apis/trending/index.ts:12]: I don't think so, PMTK may will have other data provider.

File: packages/maskbook/src/plugins/Trader/settings.ts (hunk 6 above)
Comment (guanbinrui) [reply packages/maskbook/src/plugins/Trader/settings.ts:89]: This helps throw an error for future data providers was added. For every new provider should impls those functions.

File: packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts (hunk 10 above)
Comment (guanbinrui) [reply packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts:68]: Force devs to impl this function for future data providers.

File: packages/maskbook/src/plugins/Trader/settings.ts (hunk 6 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/settings.ts:89]: got it

File: packages/maskbook/src/plugins/Trader/apis/trending/prediction_market_tokens.json [hunk 12]
```diff
@@ -0,0 +1,36 @@
+[
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/apis/trending/prediction_market_tokens.json:1]: Hi, move this file to `apis/uniswap/` folder.

File: packages/maskbook/src/plugins/Trader/apis/trending/index.ts [hunk 13]
```diff
@@ -283,9 +286,31 @@ export async function getCoinTrendingByKeyword(
     // prefer coins on the etherenum network
     const coin = coins.find((x) => x.eth_address) ?? first(coins)
     if (!coin) return null
+    if (dataProvider === DataProvider.UNISWAP) {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/apis/trending/index.ts:289]: Please move this logic to `getCoinTrendingById `.

File: packages/maskbook/src/plugins/Trader/apis/trending/prediction_market_tokens.json (hunk 12 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/apis/trending/prediction_market_tokens.json:1]: okay

File: packages/maskbook/src/plugins/Trader/apis/trending/index.ts (hunk 13 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/apis/trending/index.ts:289]: okay

File: packages/maskbook/src/plugins/Trader/apis/trending/index.ts (hunk 11 above)
Comment (guanbinrui) [reply packages/maskbook/src/plugins/Trader/apis/trending/index.ts:12]: But you treat those tokens owned by Uniswap provider. So it's better to encapsulate to `apis/uniswap`.

File: packages/maskbook/src/plugins/Trader/apis/trending/index.ts (hunk 13 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/apis/trending/index.ts:289]: when dataProvider === DataProvider.UNISWAP, does not getCoinTrending 'ById'

Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/apis/trending/index.ts:289]: I move `getCoinTrendingFromUniswap` to apis/uniswap/index.

File: packages/maskbook/src/plugins/Trader/apis/trending/index.ts (hunk 11 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/apis/trending/index.ts:12]: ok

File: packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts (hunk 10 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts]: Impl which function, resolveCoinAddress?

Comment (guanbinrui) [reply packages/maskbook/src/plugins/Trader/apis/trending/hotfix.ts]: yes.

**Review by guanbinrui — APPROVED** [review]

File: packages/maskbook/src/plugins/Trader/pipes.ts [hunk 14]
```diff
@@ -36,6 +38,8 @@ export function resolveDataProviderLink(dataProvider: DataProvider) {
             return 'https://www.coingecko.com/'
         case DataProvider.COIN_MARKET_CAP:
             return 'https://coinmarketcap.com/'
+        case DataProvider.UNISWAP:
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/pipes.ts:41]: use table driven.

```ts
const providers: Record<DataProvider, string> = {
    [DataProvider.COIN_GECKO]: '... url ...'
}
return providers[dataProvider]
```

File: packages/maskbook/src/plugins/Trader/settings.ts [hunk 15]
```diff
@@ -97,13 +100,19 @@ const coinMarketCapPreferredCoinId = createInternalSettings<string>(
     `${PLUGIN_IDENTIFIER}+currentCoinMarketCapPreferredCoinId`,
     '{}',
 )
+const coinUniswapPreferredCoinId = createInternalSettings<string>(
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/settings.ts:103]: use table driven.

Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Trader/settings.ts:103]: It's a good chance to try table driven! But this pr is urgent, will do it further.

File: packages/maskbook/src/plugins/Trader/pipes.ts (hunk 14 above)
Comment (guanbinrui) [reply packages/maskbook/src/plugins/Trader/pipes.ts:41]: @zhouhanseng If you decide to refactor this to table-driven please take `EnumRecord` a look in the codebase.

**Pull request merged**